dali/pipeline/operator/builtin/merge.cc

```diff
@@ -0,0 +1,173 @@
+  // pinnedness of the memory
+  output.set_device_id(device_id_);
+  output.set_pinned(*pinned_);
+
+  output.SetSize(input_sample_count_);
+
+  for (int output_sample_idx = 0; output_sample_idx < predicate.num_samples();
+       output_sample_idx++) {
+    int input_category_idx = get_category_index(predicate, output_sample_idx);
+    auto &input = ws.template Input<Backend>(input_category_idx);
+
+    // get the index within input category and increment for the next occurrence.
+    int input_sample_idx = sample_idx_in_input[input_category_idx];
+    sample_idx_in_input[input_category_idx]++;
+
+    if (std::is_same_v<Backend, GPUBackend> || input.is_pinned() == *pinned_) {
+      output.SetSample(output_sample_idx, input, input_sample_idx);
+    } else {
+      // Pessimistic variant, we need to copy.
+      // TODO(klecki): Do one allocation, where samples that we share are 0-volumed - this might
+      // be perf optimization reducing the number of allocations to 1.
+      CopySampleToOutput(output, output_sample_idx, input, input_sample_idx, ws);
+    }
+  }
+  FinalizeCopy(ws);
+  WriteTestsDiagnostics(ws);
+}
+
+template <>
+void Merge<CPUBackend>::CopySampleToOutput(TensorList<CPUBackend> &output, int output_sample_idx,
+                                           const TensorList<CPUBackend> &input,
+                                           int input_sample_idx, Workspace &ws) {
+  auto &tp = ws.GetThreadPool();
+  tp.AddWork(
+      [&output, &input, output_sample_idx, input_sample_idx](int thread_idx) {
+        output.ResizeSample(output_sample_idx, input.shape()[input_sample_idx]);
+        output.CopySample(output_sample_idx, input, input_sample_idx, output.order());
+      },
+      volume(input.tensor_shape_span(input_sample_idx)));
+}
+
+
+template <>
+void Merge<CPUBackend>::FinalizeCopy(Workspace &ws) {
+  ws.GetThreadPool().RunAll();
+}
+
+
+template <>
+void Merge<GPUBackend>::CopySampleToOutput(TensorList<GPUBackend> &output, int output_sample_idx,
+                                           const TensorList<GPUBackend> &input,
+                                           int input_sample_idx, Workspace &ws) {
+  assert(false && "This codepath should not be executed");
+}
+
+
+template <>
+void Merge<GPUBackend>::FinalizeCopy(Workspace &ws) {
+  assert(false && "This codepath should not be executed");
+}
+
+
+template <typename Backend>
+void Merge<Backend>::RegisterTestsDiagnostics() {
+  this->RegisterDiagnostic("input_0_pinned", &in_0_pinned_);
+  this->RegisterDiagnostic("input_1_pinned", &in_1_pinned_);
+  this->RegisterDiagnostic("output_pinned", &out_pinned_);
+}
+
+
+template <typename Backend>
+void Merge<Backend>::WriteTestsDiagnostics(const Workspace &ws) {
+  in_0_pinned_ = ws.template Input<Backend>(0).is_pinned();
+  in_1_pinned_ = ws.template Input<Backend>(1).is_pinned();
+  out_pinned_ = ws.template Output<Backend>(0).is_pinned();
+}
+
+
+DALI_SCHEMA(experimental___Merge)
```

Comment:
It doesn't matter much now because they are hidden, but I wonder if we should be more specific with the operator name. `BatchSplit` `BatchMerge` or similar. Split and Merge are names that we might want to use for some general purpose ops.

Reply:
Or even better `batch.Split` `batch.Merge`

Reply:
I can agree on the BatchSplit, but I think batch.split is a bit weird.
Still, for now they are intended to be hidden (and private).

Reply:
_conditional.[Split, Merge] would be nice, I think (as discussed offline)

Reply:
Also, `_conditional.split` was suggested.